wrappers/services.go

```diff
@@ -669,27 +671,26 @@ func (es *ensureSnapServicesContext) ensureSnapsSystemdServices() (*quota.QuotaG
 		if s.Type() == snap.TypeSnapd {
 			return nil, fmt.Errorf("internal error: adding explicit services for snapd snap is unexpected")
 		}
+		if snapSvcOpts == nil {
+			snapSvcOpts = &SnapServicesOptions{}
+		}
 
 		// always use RequireMountedSnapdSnap options from the global options
-		genServiceOpts := &AddSnapServicesOptions{
+		genServiceOpts := &generateSnapServicesOptions{
 			RequireMountedSnapdSnap: es.opts.RequireMountedSnapdSnap,
-		}
-		if snapSvcOpts != nil {
-			// and if there are per-snap options specified, use that for
-			// VitalityRank
-			genServiceOpts.VitalityRank = snapSvcOpts.VitalityRank
-			genServiceOpts.QuotaGroup = snapSvcOpts.QuotaGroup
-
-			if snapSvcOpts.QuotaGroup != nil {
-				if err := neededQuotaGrps.AddAllNecessaryGroups(snapSvcOpts.QuotaGroup); err != nil {
-					// this error can basically only be a circular reference
-					// in the quota group tree
-					return nil, err
-				}
+			VitalityRank:            snapSvcOpts.VitalityRank,
+			QuotaGroup:              snapSvcOpts.QuotaGroup,
+		}
+		if snapSvcOpts.QuotaGroup != nil {
+			if err := neededQuotaGrps.AddAllNecessaryGroups(snapSvcOpts.QuotaGroup); err != nil {
+				// this error can basically only be a circular reference
+				// in the quota group tree
+				return nil, err
 			}
 		}
 
-		if err := es.ensureSnapSystemdUnits(s, genServiceOpts); err != nil {
+		svcQuotaMap := snapSvcOpts.ServiceQuotaMap
```

Comment:
do we need to check that if this is set snapSvcOpts.QuotaGroup was also set, and the groups are all subgroups of it?

Reply:
Good suggestion! added unit tests for this case